src/languages/html.coffee

```diff
@@ -0,0 +1,561 @@
+
+    if node.nodeName is '#comment'
+      node.type = 'comment'
+      return
+
+    if not @isBlockNode node
+      node.type = 'emptyTag'
+      if node.__location?
+        @setAttribs node, @text[node.__location.start...node.__location.end]
+      return
+
+    node.type = 'blockTag'
+
+    node.__indentLocation = {start: node.__location.startTag.end}
+
+    if node.childNodes?.length > 0
+      node.__indentLocation.end = node.childNodes[node.childNodes.length - 1].__location.end
+    else
+      node.__indentLocation.end = node.__location.startTag.end
+
+    if not node.__location.endTag
+      node.__location.end = node.__indentLocation.end
+
+    @setAttribs node, @text[node.__location.start...node.__indentLocation.start]
+
+  makeIndentBounds: (node) ->
+    bounds = {
+      start: @positions[node.__indentLocation.start]
+      end: @positions[node.__indentLocation.end]
+    }
+
+    if node.__location.endTag?
+      lastLine = @positions[node.__location.endTag.start].line - 1
+      if lastLine > bounds.end.line or (lastLine is bounds.end.line and @lines[lastLine].length > bounds.end.column)
+        bounds.end = {
+          line: lastLine
+          column: @lines[lastLine].length
+        }
+      else if node.__indentLocation.start is node.__indentLocation.end and node.__location.endTag
+        bounds.end = @positions[node.__location.endTag.start]
+
+    return bounds
+
+  trimText: (node) ->
+    text = node.value
+    location = node.__location
+    text = text.split '\n'
+    i = 0
+    while text[i].trim().length is 0
+      location.start += text[i].length + 1
+      i++
+    j = text.length - 1
+    while text[j].trim().length is 0
+      j--
+    text = text[i..j].join '\n'
+    location.end = location.start + text.length
+    if i isnt 0
+      leftTrimText = text.trimLeft()
+      location.start += text.length - leftTrimText.length
+      text = leftTrimText
+    node.value = text
+
+  getSocketLevel: (node) -> helper.ANY_DROP
+
+  htmlBlock: (node, depth, bounds) ->
+    @addBlock
+      bounds: bounds ? @getBounds node
+      depth: depth
+      precedence: @getPrecedence node
+      color: @getColor node
+      classes: @getClasses node
+      socketLevel: @getSocketLevel node
+
+  htmlSocket: (node, depth, precedence, bounds, classes) ->
+    @addSocket
+      bounds: bounds ? @getBounds node
+      depth: depth
+      precedence: precedence
+      classes: classes ? @getClasses node
+
+  getIndentPrefix: (bounds, indentDepth, depth) ->
+    if bounds.end.line - bounds.start.line < 1
+      return DEFAULT_INDENT_DEPTH
+    else
+      tmp = 1
+      while @lines[bounds.start.line + tmp].trim().length is 0
+        tmp++
+      line = @lines[bounds.start.line + tmp]
+      lineIndent = line[indentDepth...(line.length - line.trimLeft().length)]
+      if lineIndent.length >= DEFAULT_INDENT_DEPTH.length
+        return lineIndent
+      else
+        return DEFAULT_INDENT_DEPTH
+
+  markRoot: ->
+    @positions = []
```

Comment:
question for @sakagg and @dabbler0 - when parsing within a specific context, such as within <script>, can it parse with different rules?

For example: `<script>document.write('<img>');</script>` - here, parsers treat everything inside as text even though there is an `<img>`.  If editing a socket inside a `<script>`, will it reparse in script context corrrectly?